Adding support for ⚡ Groq API

someshfengde opened the pull request:
##  Now we can use Groq API within dspy.
Makes querying to LLMs lot faster. 💨 

### Thanks @ivarflakstad for helping me debug.

## Usage: 

```
import dspy
groq_model = dspy.GROQ(api_key ="gsk_***" )
dspy.settings.configure(lm = groq_model)
class BasicQA(dspy.Signature):
     "hi mom"
     question = dspy.InputField()
     answer = dspy.OutputField(desc="often between 1 and 5 words")
 
generate_answer = dspy.Predict(BasicQA)
pred = generate_answer(question="what is the meaning of life")
print(pred)
```
Output 
```
Prediction(
    answer='Meaning varies person to person.'
)
```

ivarflakstad commented on dsp/modules/groq_client.py:
```diff
@@ -0,0 +1,166 @@
+        assert return_sorted is False, "for now"
+        response = self.request(prompt, **kwargs)
+
+        if dsp.settings.log_openai_usage:
```
If we are to use this field in this context it should probably be renamed to something more generic. Like `log_third_party_usage` or `log_external_usage`?
What do you think @okhat @arnavsinghvi11 ?
(Tagging you guys since you authored the settings and the field respectively ;) )

insop commented:
@ivarflakstad thank you for the PR.
Could you add the documentation at docs/api/language_model_clients

ivarflakstad commented:
@someshfengde deserves the thanks😊

someshfengde commented:
yes sure @insop  will do it. :)

someshfengde commented:
Hi @insop  I've added the documentation for the same :)

insop commented on dsp/modules/lm.py:
```diff
@@ -82,6 +82,9 @@ def inspect_history(self, n: int = 1, skip: int = 0):
             elif provider == "clarifai" or provider == "claude" :
                 text=choices
+            elif provider == "groq":
+                # print(choices)
```
remove the commented line?

someshfengde replied:
done

insop commented:
Thank you @someshfengde , looks good to me.

someshfengde commented:
@insop  removed comment also resolved merge conflicts

ebudmada commented:
Very great, can't handle myself the see the mistral8x7b tests. Groq is very cheap, can it be 0,27$ for million tokens?! Great great great ! Thank you @someshfengde

pfa34488 commented:
When I run the script. I get dspy cannot find GROQ.  I pip install dspy-ai. Am I missing anything?

ebudmada commented:
@pfa34488 Yes you have to integrate the changes yourself! This workflow is waiting approval

pfa34488 commented:
Thanks so much! I see the module at curieo.org/dspy under dsp modules; I see from .groq_client import * and the [groq_client.py](https://github.com/curieo-org/dspy/blob/initial-groq-support/dsp/modules/groq_client.py) file. I am not sure how to add. Do i find the pip install dspy-ai or git clone (https://github.com/curieo-org/dspy.git) or something else? Please help!

ebudmada commented:
Yes there is 8 file change in this PR, do it yourself or wait until it is implemented

okhat commented:
Thanks so much. Checks fail because this has imports that are always on but are only needed for those that use this class. Please see how cohere and many other modules do this, for a different pattern: import on demand.

pfa34488 commented:
I got it LoL. You guys are sharp. From Newbe

someshfengde commented:
Hi @pfa34488 you can clone a branch from repo with branch initial-groq-support. 

Sorry currently you can't work with the main branch I'm resolving an error for The Together api over main.

someshfengde commented:
> Very great, can't handle myself the see the mistral8x7b tests. Groq is very cheap, can it be 0,27$ for million tokens?! Great great great ! Thank you @someshfengde

Welcome 🤗

someshfengde commented:
> When I run the script. I get dspy cannot find GROQ.  I pip install dspy-ai. Am I missing anything?
> 

It's still not integrated into the main branch. It'll soon be integrated 🙂

someshfengde commented:
> Thanks so much. Checks fail because this has imports that are always on but are only needed for those that use this class. Please see how cohere and many other modules do this, for a different pattern: import on demand. 

Yes sure will do it 👍🏻

someshfengde commented:
> Thanks so much. Checks fail because this has imports that are always on but are only needed for those that use this class. Please see how cohere and many other modules do this, for a different pattern: import on demand.

I've removed the unused modules and ran `ruff check . --fix-only`.

arnavsinghvi11 commented:
Hi @someshfengde, thanks for removing the other imports. @okhat is actually referring to following import on-demand behavior ([as done in Cohere](https://github.com/stanfordnlp/dspy/blob/9ca1ce53eeada380a6f786951198946f3a239c72/dsp/modules/cohere.py#L8)) to only import libraries needed for this class. Could you remove those imports from the pyproject.toml as well and add instructions to the documentation one how to install them locally ([Cohere documentation for reference](https://github.com/stanfordnlp/dspy/blob/9ca1ce53eeada380a6f786951198946f3a239c72/docs/docs/deep-dive/language_model_clients/remote_models/Cohere.mdx#L8))

someshfengde commented:
Hi @arnavsinghvi11  thanks for the suggestion. It's done I've already added the docs previously :)

okhat commented on pyproject.toml:
```diff
@@ -108,6 +109,7 @@ sphinx_rtd_theme = { version = "*", optional = true }
 autodoc_pydantic = { version = "*", optional = true }
 sphinx-reredirects = { version = "^0.1.2", optional = true }
 sphinx-automodapi = { version = "0.16.0", optional = true }
+groq = "^0.4.2"
```
Shouldn't this be optional? We don't want a dependency on groq

arnavsinghvi11 replied:
@someshfengde adding `optional = true` as done with the other dependencies should resolve this

someshfengde replied:
Hi @arnavsinghvi11  @okhat  I've added it as optional

someshfengde replied:
Also resolved the merge conflicts.

arnavsinghvi11 replied:
Thanks @someshfengde ! are the poetry.lock changes needed for the PR as well? not sure if they are. will merge following this confirmation!

someshfengde replied:
👍

someshfengde commented:
@okhat any updates on this? if there is something wrong please let me know.

someshfengde commented:
Hi @okhat  merged changes and resolved conflicts again

arnavsinghvi11 commented:
Hi @someshfengde , the PR should be good to go but I still don't see the need for the poetry.lock changes? It seems like there are a lot of extra changes not relevant to the DSPy library, but please let me know if I'm misinterpreting. Will merge following your response/changes.

someshfengde commented:
> Hi @someshfengde , the PR should be good to go but I still don't see the need for the poetry.lock changes? It seems like there are a lot of extra changes not relevant to the DSPy library, but please let me know if I'm misinterpreting. Will merge following your response/changes. 

Hi the poetry lock is auto generated from the poetry. Which is dependent on pyproject.toml I don't think that should be any problem?

someshfengde commented:
> > Hi @someshfengde , the PR should be good to go but I still don't see the need for the poetry.lock changes? It seems like there are a lot of extra changes not relevant to the DSPy library, but please let me know if I'm misinterpreting. Will merge following your response/changes. 
> 
> Hi the poetry lock is auto generated from the poetry. Which is dependent on pyproject.toml I don't think that should be any problem?

And documentation says we are not supposed to edit the poetry.lock file by hand let me know what needs to be done?

arnavsinghvi11 commented:
> > Hi @someshfengde , the PR should be good to go but I still don't see the need for the poetry.lock changes? It seems like there are a lot of extra changes not relevant to the DSPy library, but please let me know if I'm misinterpreting. Will merge following your response/changes.
> 
> Hi the poetry lock is auto generated from the poetry. Which is dependent on pyproject.toml I don't think that should be any problem?

Ideally, we don't want to commit the poetry.lock to the library itself and keep it auto-generated locally for the user. If the existing PR feature functions without these changes, we can remove the poetry.lock file and merge without it. Let me know if that makes sense

someshfengde commented:
> > > Hi @someshfengde , the PR should be good to go but I still don't see the need for the poetry.lock changes? It seems like there are a lot of extra changes not relevant to the DSPy library, but please let me know if I'm misinterpreting. Will merge following your response/changes.
> > 
> > 
> > Hi the poetry lock is auto generated from the poetry. Which is dependent on pyproject.toml I don't think that should be any problem?
> 
> Ideally, we don't want to commit the poetry.lock to the library itself and keep it auto-generated locally for the user. If the existing PR feature functions without these changes, we can remove the poetry.lock file and merge without it. Let me know if that makes sense

Yes it does. I've removed the file

arnavsinghvi11 commented:
@someshfengde ah I meant, just remove the changes from your poetry.lock file and keep the current version as-is. Apologies for the back-and-forth!

someshfengde commented:
> @someshfengde ah I meant, just remove the changes from your poetry.lock file and keep the current version as-is. Apologies for the back-and-forth!

Hi @arnavsinghvi11 I've made poetry lock same as the main branch. Sorry and thanks for your patience too :)

arnavsinghvi11 commented on .gitignore:
```diff
@@ -46,3 +46,8 @@ finetuning_ckpts/
 *.log
 *.db
+.venv
+temp_req.py
```
can you remove these extra files here? we don't want to include extraneous files in the library's gitignore

arnavsinghvi11 commented:
Hi @someshfengde , the CI tests seem to still fail. Could you run these following commands to fix it?


Run Tests failure: 
`Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run `poetry lock [--no-update]` to fix it.`

Check Ruff Fix (formatting) failure:
`ruff check . --fix-only`


Build Poetry failure:
```
Run python -c "import dspy"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/runner/work/dspy/dspy/dspy/__init__.py", line 1, in <module>
    import dsp
  File "/home/runner/work/dspy/dspy/dsp/__init__.py", line 1, in <module>
    from .modules import *
  File "/home/runner/work/dspy/dspy/dsp/modules/__init__.py", line [12](https://github.com/stanfordnlp/dspy/actions/runs/8431136781/job/23113260978?pr=637#step:8:13), in <module>
    from .together_client import *
  File "/home/runner/work/dspy/dspy/dsp/modules/together_client.py", line 4, in <module>
    from typing import Any, Literal, Optional, Required
ImportError: cannot import name 'Required' from 'typing' (/opt/hostedtoolcache/Python/3.9.[18](https://github.com/stanfordnlp/dspy/actions/runs/8431136781/job/23113260978?pr=637#step:8:19)/x64/lib/python3.9/typing.py)
Error: Process completed with exit code 1.
```

arnavsinghvi11 commented:
I've also left another comment for removing the extraneous files from .gitignore

arnavsinghvi11 commented on dsp/modules/hf_client.py:
```diff
@@ -270,37 +268,48 @@ def _generate(self, prompt, use_chat_api=False, **kwargs):
 
-        try:
             with self.session.post(url, headers=headers, json=body) as resp:
+                print(resp)
```
can you remove this print statement?

arnavsinghvi11 commented on dsp/modules/hf_client.py:
```diff
@@ -270,37 +268,48 @@ def _generate(self, prompt, use_chat_api=False, **kwargs):
             with self.session.post(url, headers=headers, json=body) as resp:
+                print(resp)
                 resp_json = resp.json()
+                print(resp_json)
```
can you remove this print statement?

arnavsinghvi11 commented:
Also, can you remove all the related Together changes? seems they have been added in some extra commits I didn't see before but they are unrelated to the Groq LM in this PR and are modifying the integrated dspy.Together LM. Let me know if you have any questions about this (but these changes could possibly be impacting the test failures too!)

arnavsinghvi11 commented on dspy/__init__.py:
```diff
@@ -21,12 +21,13 @@
 
 Anyscale = dsp.Anyscale
-Together = dsp.Together
+Together = dsp.TogetherLM
```
this points to the Together client found in hf_client

arnavsinghvi11 commented on dsp/modules/together_client.py:
```diff
@@ -0,0 +1,171 @@
+import logging
```
DSPy already supports a Together client via hf_client. this file (and related changes can be removed)

someshfengde commented:
Hi @arnavsinghvi11  Removed the together related files, removed the gitignore added files too. Updated the poetry lock with poetry lock --no-update 
and ran ruff check . --fix-only too 

Thanks for the review

arnavsinghvi11 commented:
Hi @someshfengde , the tests pass but there are still conflicts. Can you check if its a merge conflict on your end?

if not, I will checkout and merge the changes to main through command line.

someshfengde commented:
> Hi @someshfengde , the tests pass but there are still conflicts. Can you check if its a merge conflict on your end?
> 
> if not, I will checkout and merge the changes to main through command line.

it was poetry lock conflict for hash I resolved it. if someone push new commit that hash is different causing conflict in PR

okhat commented:
Happy to merge here once conflicts are gone

someshfengde commented:
Hi for every commit in poetry lock file will cause the merge conflict I've been resolving these conflicts lot's of times already 😢

okhat commented:
Ah got it. Why not just remove/revert the lock and other files you don’t need to change? You just need to change like 4-5 files

someshfengde commented:
> Ah got it. Why not just remove/revert the lock and other files you don’t need to change? You just need to change like 4-5 files

yes just a sec will do it rn

someshfengde commented:
@okhat  it's done

arnavsinghvi11 commented:
Hi @someshfengde , it's still failing the poetry tests. 

Can you run `poetry lock --no-update` and `poetry install --no-interaction` again?

someshfengde commented:
> Hi @someshfengde , it's still failing the poetry tests.
> 
> Can you run `poetry lock --no-update` and `poetry install --no-interaction` again?

I've ran this many times we are falling behind since every new push to the main repo will cause merge issues in this one?

arnavsinghvi11 commented:
After merging and running the commands, are there any updates you need to commit? seems like that should fix it or removing the poetry.lock which has extraneous changes.

someshfengde commented:
@arnavsinghvi11  I've fixed the poetry lock for now

arnavsinghvi11 commented:
Closing since #773 is merged. Thanks again @someshfengde !
